active/NumberBox/NumberBox.md

```diff
@@ -0,0 +1,234 @@
+only when there's a bug in the caller, such as argument exception.  But if for some
+reason it's necessary for a caller to catch an exception from an API, call that
+out with an explanation either here or in the Examples -->
+
+## API Notes
+<!-- Option 1: Give a one or two line description of each API (type
+and member), or at least the ones that aren't obvious
+from their name.  These descriptions are what show up
+in IntelliSense. For properties, specify the default value of the property if it
+isn't the type's default (for example an int-typed property that doesn't default to zero.) -->
+
+<!-- Option 2: Put these descriptions in the below API Details section,
+with a "///" comment above the member or type. -->
+
+
+## API Notes
+
+### Notable Properties  
+
+| Name | Description |
+|:-:|:--|
+| StepFrequency | Gets or sets the value part of a value range that steps should be created for. |
+
+## API Details
+<!-- The exact API, in MIDL3 format (https://docs.microsoft.com/en-us/uwp/midl-3/) -->
+
+```c++ 
+runtimeclass NumberBoxValueChangedEventArgs
+{
+    String Text;
+    Double Value
+};
+
+runtimeclass NumberBoxTextChangedEventArgs
+{
+    String Text;
+    Double Value
+};
+
+unsealed runtimeclass NumberBox : Windows.UI.Xaml.Controls.TextBox
+{
+    NumberBox();
+    
+    Double Value;
+    Boolean AcceptsCalculation;
+    
+    Boolean UpDownButtonsEnabled;
+    Boolean HyperDragEnabled;
+    Boolean HyperScrollEnabled;
+    Double StepFrequency;
+  
+    Boolean AreLeadingZerosTrimmed;
+    Double DecimalPrecision;
+    Boolean DoesInputRound;
+    
+    Double MinValue;
+    Double MaxValue;
+    
+    Boolean IsInvalidInputOverwritten
+    
+
+    NumberBoxTemplateSettings TemplateSettings{ get; };
+
+    event Windows.Foundation.TypedEventHandler<NumberBox, NumberBoxValueChangedEventArgs> ValueChanged;
+    event Windows.Foundation.TypedEventHandler<NumberBox, NumberBoxTextChangedEventArgs> TextChanged;
+
+    static Windows.UI.Xaml.DependencyProperty ValueProperty{ get; };
+    static Windows.UI.Xaml.DependencyProperty AcceptsCalculationProperty{ get; };
+    
+    static Windows.UI.Xaml.DependencyProperty UpDownButtonsEnabledProperty{ get; };
+    static Windows.UI.Xaml.DependencyProperty HyperDragEnabledProperty{ get; };
+    static Windows.UI.Xaml.DependencyProperty HyperScrollEnabledProperty{ get; };   
+    static Windows.UI.Xaml.DependencyProperty StepFrequencyProperty{ get; };
+    
+    static Windows.UI.Xaml.DependencyProperty AreLeadingZerosTrimmedProperty{ get; };
+    static Windows.UI.Xaml.DependencyProperty DecimalPrecisionProperty{ get; };
+    static Windows.UI.Xaml.DependencyProperty DoesInputRoundProperty{ get; };
+    
+    static Windows.UI.Xaml.DependencyProperty MinValueProperty{ get; };
+    static Windows.UI.Xaml.DependencyProperty MaxValueProperty{ get; };
+    
+    static Windows.UI.Xaml.DependencyProperty IsInvalidInputOverwrittenProperty{ get; };
+}
+```
+
+## Appendix
```

Comment:
The appendix is probably an appropriate place to discuss keyboard/gamepad behavior. Of note, I'm curious how the UP/DOWN buttons work, as well as the ESC key. I believe ENTER/TAB is discussed elsewhere, but it'd be worthwhile to condense that info into a single place.

Reply:
Agreed. I believe @Austin-Lamb has been advocating that those sections be part of the standard, required template and I intend to include it shortly.

Will leave this open until resolved.

Reply:
My gut tells me **Esc**ape should cancel a change in the value that was already there.  It does not in Photoshop, with annoys me all the time I wish to cancel a change I had already started making.

Reply:
**Esc**ape already has window level functionality at some levels such as cancelling/closing a dialog.
If the NumberBox was inside such a dialog it could give the impression that the whole dialog wasn't working as expected.
Undo (ctrl+Z) allows reverting changes that have been made.

Reply:
> Escape already has window level functionality at some levels such as cancelling/closing a dialog.

So does ENTER. We have some precedent that ESC will "reset" the field - the biggest one that comes to mind is ComboBox. 

I think ENTER and ESC go hand-in-hand in that if NumberBox responds to one, it should respond to both.

Reply:
> I think ENTER and ESC go hand-in-hand in that if NumberBox responds to one, it should respond to both.

I disagree, I think Enter and Escape are distinct concepts. Escaping out of a TextBox or NumberBox doesn't conceptually make sense. In ComboBox, the Esc key behavior varies depending on whether the ComboBox's flyout is open or closed. On a closed ComboBox, Escape doesn't do anything.

Reply:
You would be **Esc**aping from the calculation that is about to be made in the textbox, and removing focus from it.  **Enter** would be submitting the calculation and accepting the result it gives as the new Value for the control.

Reply:
Escape and undo aren't exactly the same concept. They're related, more closely in my opinion than Escape and Enter, but still distinct. We tend to equate escape with back, at least in flyout scenarios. Matt had mentioned Ctrl+Z for undo (notably, not the same as **Esc**ape), but even that, to me, feels outside the scope of NumberBox. I don't believe that the default TextBox, or any other Xaml control, has the concept of history/clipboard built in.

Reply:
> Escape and undo aren't exactly the same concept. They're related, more closely in my opinion than Escape and Enter, but still distinct. We tend to equate escape with back, at least in flyout scenarios. Matt had mentioned Ctrl+Z for undo (notably, not the same as **Esc**ape), but even that, to me, feels outside the scope of NumberBox. I don't believe that the default TextBox, or any other Xaml control, has the concept of history/clipboard built in.

Just to clarify, I am not saying Escape would undo the last calculation.

If you enter in some text like: **128 / 2**, but before you actually press Enter to actually do the calculation, and replace the Value and Text with the calculated value - pressing Escape would remove the text you have entered, will not do the calculation, and will replace the Text with the Value property, putting the value back the way it was before you started typing.

Reply:
I'll leave it to @SavoySchuler to spec, but to me this doesn't quite feel like the right use of Escape, and I think that our [Esc key documentation](https://docs.microsoft.com/en-us/windows/uwp/design/input/keyboard-interactions#enter-key) supports this point of view: *"The Esc key lets a user cancel transient UI (along with any ongoing actions in that UI)."*

Reply:
I believe the Calculation you type into the Text property is Transient until it is confirmed by pressing Enter, or clicking off the NumberBox after entering a valid calculation string.  There will possibly be a preview value displayed whilst the NumberBox still has focus, but having the ability to back out of the calculation, would be useful.

Reply:
And in general, single-line text boxes don't eat the Enter key. That's reserved for form submission or dialog confirmation and we should be careful about this control getting in the way of that.

Reply:
Thanks for building up a pro/con case for each! It appears that that behaviors that may be desirable from the user perspective may not align with broad expectations our ecosystem has created through legacy. I will have both Escape and Enter reviewed today by the broader team shortly.

Reply:
Resolving conversation: loss of focus, "Enter", and stepping will trigger evaluation, as noted below in review summary and in spec.